eth/tracers/api.go

```diff
@@ -862,71 +862,78 @@ func (api *API) TraceCall(ctx context.Context, args ethapi.TransactionArgs, bloc
 	)
-	switch {
-	case config == nil:
-		tracer = logger.NewStructLogger(nil)
-	case config.Tracer != nil:
-		// Define a meaningful timeout of a single transaction trace
-		timeout := defaultTraceTimeout
-		if config.Timeout != nil {
-			if timeout, err = time.ParseDuration(*config.Timeout); err != nil {
-				return nil, err
-			}
-		}
-		if t, err := New(*config.Tracer, txctx); err != nil {
+	// Use struct logger
+	if config == nil || config.Tracer == nil {
+		return api.logTx(message, txctx, vmctx, statedb, config, txContext)
+	}
+
+	// Define a meaningful timeout of a single transaction trace
+	timeout := defaultTraceTimeout
+	if config.Timeout != nil {
+		if timeout, err = time.ParseDuration(*config.Timeout); err != nil {
 			return nil, err
-		} else {
-			deadlineCtx, cancel := context.WithTimeout(ctx, timeout)
-			go func() {
-				<-deadlineCtx.Done()
-				if errors.Is(deadlineCtx.Err(), context.DeadlineExceeded) {
-					t.Stop(errors.New("execution timeout"))
-				}
-			}()
-			defer cancel()
-			tracer = t
 		}
-	default:
-		tracer = logger.NewStructLogger(config.Config)
 	}
+	tracer, err := New(*config.Tracer, txctx)
+	if err != nil {
+		return nil, err
+	} else {
+		deadlineCtx, cancel := context.WithTimeout(ctx, timeout)
+		go func() {
+			<-deadlineCtx.Done()
+			if errors.Is(deadlineCtx.Err(), context.DeadlineExceeded) {
+				tracer.Stop(errors.New("execution timeout"))
+			}
+		}()
+		defer cancel()
+	}
+
 	// Run the transaction with tracing enabled.
 	vmenv := vm.NewEVM(vmctx, txContext, statedb, api.backend.ChainConfig(), vm.Config{Debug: true, Tracer: tracer, NoBaseFee: true})
-
 	// Call Prepare to clear out the statedb access list
 	statedb.Prepare(txctx.TxHash, txctx.TxIndex)
-
-	result, err := core.ApplyMessage(vmenv, message, new(core.GasPool).AddGas(message.Gas()))
+	_, err = core.ApplyMessage(vmenv, message, new(core.GasPool).AddGas(message.Gas()))
 	if err != nil {
 		return nil, fmt.Errorf("tracing failed: %w", err)
 	}
 
-	// Depending on the tracer type, format and return the output.
-	switch tracer := tracer.(type) {
-	case *logger.StructLogger:
-		// If the result contains a revert reason, return it.
-		returnVal := fmt.Sprintf("%x", result.Return())
-		if len(result.Revert()) > 0 {
-			returnVal = fmt.Sprintf("%x", result.Revert())
-		}
-		return &ethapi.ExecutionResult{
-			Gas:         result.UsedGas,
-			Failed:      result.Failed(),
-			ReturnValue: returnVal,
-			StructLogs:  ethapi.FormatLogs(tracer.StructLogs()),
-		}, nil
+	return tracer.GetResult()
+}
+
+// logTx executes the given message and logs the steps of the execution
+// using a StructLogger.
+func (api *API) logTx(message core.Message, txctx *Context, vmctx vm.BlockContext, statedb *state.StateDB, config *TraceConfig, txContext vm.TxContext) (*ethapi.ExecutionResult, error) {
+	var tracer *logger.StructLogger
+	if config == nil {
+		tracer = logger.NewStructLogger(nil)
+	} else {
+		tracer = logger.NewStructLogger(config.Config)
+	}
```

Comment:
As it is, you now have this config==nil check both here and in the caller, `traceTx`. Maybe you can just pass the `tracer` to this method, and do this instantiation in `traceTx` (shave off a few lines)